cmd/dra-example-kubeletplugin/state.go

```diff
@@ -173,3 +210,89 @@ func (s *DeviceState) prepareDevices(claim *resourceapi.ResourceClaim) (Prepared
 func (s *DeviceState) unprepareDevices(claimUID string, devices PreparedDevices) error {
 	return nil
 }
+
+func (p *PreparedDevice) applyConfig() error {
+	return nil
+}
+
+// GetDevices extracts the list of drapbv1.Devices from PreparedDevices.
+func (pds PreparedDevices) GetDevices() []*drapbv1.Device {
+	var devices []*drapbv1.Device
+	for _, pd := range pds {
+		devices = append(devices, &pd.Device)
+	}
+	return devices
+}
+
+// ptrImplementsRuntimeObject is a helper interface for GetMergedOpaqueDeviceConfig.
+//
+// It enforces that a pointer to the type parameter passed to
+// GetMergedOpaqueDeviceConfig implements the runtime.Object interface as well
+// as the GroupVersionKind() function (both of which are true if it has an
+// embedded 'metav1.TypeMeta' field).
+type ptrImplementsRuntimeObject[T any] interface {
+	*T
+	runtime.Object
+	GroupVersionKind() schema.GroupVersionKind
+}
+
+// GetMergedOpaqueDeviceConfig merges all opaque configs specified for a device request in a resource claim.
+//
+// Such configs can either be attached to the resource claim itself or to the
+// device class associated with the request. Configs found directly in the
+// resource claim take precedence over configs found in the device class.
+// Moreover, configs found earlier in the list of configs attached to either a
+// resource claim or a device class take precedence over configs found later in
+// their corresponding list.
+//
+// When merging, a config with higher precedence will overwrite fields of a
+// config with lower precedence. The only exception being that zero-valued
+// fields of higher precedence configs will *not* overwrite non-zero valued
+// fields of lower precedence configs.
+//
+// As such, the final config returned by this function will contain the merged
+// values of all non-zero fields across each of the possible configs in
+// precedence order.
+func GetMergedOpaqueDeviceConfig[T any, TI ptrImplementsRuntimeObject[T]](
+	decoder runtime.Decoder,
+	driverName string,
+	request string,
+	resourceClaim *resourceapi.ResourceClaim,
+	deviceClasses map[string]*resourceapi.DeviceClass,
+) (T, error) {
+	// Collect all configs in order of reverse precedence.
+	var allConfigs []resourceapi.DeviceConfiguration
+	for _, config := range deviceClasses[request].Spec.Config {
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+	for _, config := range resourceClaim.Spec.Devices.Config {
+		if !slices.Contains(config.Requests, request) {
+			continue
+		}
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+
+	// Merge all configs from the allConfigs slice in the order they were
+	// added. Fields in a config that is decoded earlier will be
+	// overwritten by (non-zero) fields in a config that is decoded later.
```

Comment:
I am a bit worried here that this enters "undefined behavior" territory.

The current "encoding/json" de marshaling might work like that, but I am not sure whether that is guaranteed and whether the Kubernetes apimachinery makes similar guarantees.

Reply:
This has now been changed to "replace" rather than "merge" semantics in the newest version.